lib/repositories/transaction_repository.dart

```diff
@@ -0,0 +1,137 @@
+import 'dart:async';
+
+import 'package:cloud_firestore/cloud_firestore.dart';
+import 'package:get/get.dart';
+import 'package:pispapp/controllers/ephemeral/lookup_payee_controller.dart';
+import 'package:pispapp/controllers/ephemeral/payment_details_controller.dart';
+import 'package:pispapp/controllers/ephemeral/payment_verdict_controller.dart';
+import 'package:pispapp/models/account.dart';
+import 'package:pispapp/models/transaction.dart' as transaction_model;
+
+import 'package:pispapp/repositories/interfaces/i_transaction_repository.dart';
+
+class TransactionRepository implements ITransactionRepository {
+  final Firestore firestore = Firestore.instance;
+  CollectionReference get transactions => firestore.collection('transactions');
+
+  @override
+  Future<List<transaction_model.Transaction>> getTransactions(
+      String userId, String sourceAccountId) async {
+    final userTransactions = await transactions
+        .where('userId', isEqualTo: userId)
+        .where('sourceAccountId', isEqualTo: sourceAccountId)
+        .getDocuments();
+    final List<transaction_model.Transaction> transactionList = userTransactions
+        .documents
+        .map((DocumentSnapshot element) =>
+            transaction_model.Transaction.fromJson(element.data))
+        .toList();
+    return transactionList;
+  }
+
+  @override
+  Future<String> initiateTransaction(
+      String userId, String phoneIsoCode, String phoneNumber) async {
+    final documentReference = await transactions.add(
+      <String, dynamic>{
+        'userId': userId,
+        'payee': {
+          'partyIdInfo': {
+            'partyIdType': 'MSISDN',
+            'partyIdentifier': phoneIsoCode + phoneNumber,
+          }
+        }
+      },
+    );
+    return documentReference.documentID;
+  }
+
+  @override
+  StreamSubscription<DocumentSnapshot> setupTransactionStream(
+      String transactionid) {
+    final transactionStreamSubscription =
+        transactions.document(transactionid).snapshots().listen(
```

Comment:
the mobile app should only listen to transaction documents with its own `userId`

Reply:
initiateTransaction() is always called before this and this function is always called with the returned transactionId. Does that make sense ? Or should we not make such assumptions ?

Reply:
I think you had some misunderstanding here...
The `transactions` collections in Firebase contains all of the transactions from all users, but each user should have access only for their own transaction (cannot see other users' transaction). That's why when listening for the snapshots here (e.g., creation of new document, update of a document), the user needs to pass in a filtering like `where("userId", "==", ...)`. Well, if the Firebase rules are set properly, the user shouldn't have access anyway, but let's try to be more cautious here by also setting up the filtering. You may refer to the sequence diagram [here](https://github.com/mojaloop/pisp-demo-server/blob/master/docs/design/transfer.md#1-create-listener).

Reply:
Yes that makes sense. What I am saying is, `setupTransactionStream` is always called on transactions that the current user creates. So there is no need of `"userId", "==", "..."`.